src/ControlSystem/Averager.cpp

```diff
@@ -0,0 +1,163 @@
+    ERROR(
+        "The specified avg_timescale_fac, the fraction of the damping "
+        "timescale over which averaging is performed, ("
+        << avg_tscale_fac_ << ") must be positive");
+  }
+
+  // initialize the remaining member vars
+  clear();
+}
+
+template <size_t DerivOrder>
+std::array<DataVector, DerivOrder + 1> Averager<DerivOrder>::operator()(
+    const double time) const noexcept {
+  if (not have_valid_data(time)) {
+    ERROR(
+        "Cannot return averaged values because the averager does not have "
+        "up-to-date information.");
+  }
+
+  return averaged_values_;
+}
+
+template <size_t DerivOrder>
+void Averager<DerivOrder>::clear() noexcept {
+  averaged_values_ = make_array<DerivOrder + 1>(DataVector());
+  times_.clear();
+  raw_qs_.clear();
+  weight_k_ = {0.0};
+  tau_k_ = {0.0};
+}
+
+template <size_t DerivOrder>
+void Averager<DerivOrder>::update(const double time, const DataVector& raw_q,
+                                  const DataVector& timescales) noexcept {
+  if (not raw_qs_.empty()) {
+    // ensure that the number of components passed to update is the
+    // same as the number of components in the raw_q history
+    if (raw_q.size() != raw_qs_[0].size()) {
+      ERROR("The number of components in the raw_q provided ("
+            << raw_q.size()
+            << ") does not match the size of previously supplied raw_q ("
+            << raw_qs_[0].size() << ").");
+    }
+  } else {
+    // This is the first call to update: initialize averaged values, weights and
+    // effective time (with proper number of components)
+    averaged_values_ =
+        make_array<DerivOrder + 1>(DataVector(raw_q.size(), 0.0));
+    weight_k_ = averaged_values_[0];
+    tau_k_ = averaged_values_[0];
+  }
+
+  // update deques
+  times_.emplace_front(time);
+  raw_qs_.emplace_front(raw_q);
+  if (times_.size() > DerivOrder + 1) {
+    // get rid of old data once we have a sufficient number of points
+    times_.pop_back();
+    raw_qs_.pop_back();
+  }
+
+  if (times_.size() > DerivOrder) {
+    // we now have enough data points to begin averaging
+    const DataVector tau_avg = timescales * avg_tscale_fac_;
+    const double tau_m = times_[0] - times_[1];
+
+    // update the weights and effective time
+    const DataVector old_weight{std::move(weight_k_)};
+    weight_k_ = (tau_m + old_weight) * tau_avg / (tau_m + tau_avg);
+    tau_k_ = (time * tau_m + tau_k_ * old_weight) * tau_avg /
+             (weight_k_ * (tau_m + tau_avg));
+
+    std::array<DataVector, DerivOrder + 1> raw_derivs = get_derivs();
+    for (size_t i = 0; i < averaged_values_.size(); i++) {
+      gsl::at(averaged_values_, i) =
+          (gsl::at(raw_derivs, i) * tau_m +
+           gsl::at(averaged_values_, i) * old_weight) *
+          tau_avg / (weight_k_ * (tau_m + tau_avg));
+    }
+    // replace averaged function value with raw value if
+    // not using `average_0th_deriv_of_q`
+    if (not average_0th_deriv_of_q_) {
```

Comment:
Did you average something, then check if you needed to average it, and if not overwrite it? :(

Reply:
:/ yeah, sorry about that